pkg/kds/server/server_test.go

```diff
@@ -0,0 +1,628 @@
+			Match: map[string]string{
+				"service": "*",
+			},
+		}},
+		Destinations: []*mesh_proto.Selector{{
+			Match: map[string]string{
+				"service": "*",
+			},
+		}},
+		Conf: &mesh_proto.TrafficLog_Conf{
+			Backend: "logging-backend",
+		},
+	}
+	tp1 = mesh_proto.TrafficPermission{
+		Sources: []*mesh_proto.Selector{{
+			Match: map[string]string{
+				"service": "*",
+			},
+		}},
+		Destinations: []*mesh_proto.Selector{{
+			Match: map[string]string{
+				"service": "*",
+			},
+		}},
+	}
+	tr1 = mesh_proto.TrafficRoute{
+		Sources: []*mesh_proto.Selector{{
+			Match: map[string]string{
+				"service": "*",
+			},
+		}},
+		Destinations: []*mesh_proto.Selector{{
+			Match: map[string]string{
+				"service": "*",
+			},
+		}},
+		Conf: []*mesh_proto.TrafficRoute_WeightedDestination{{
+			Weight: 10,
+			Destination: map[string]string{
+				"version": "v2",
+			},
+		}},
+	}
+	tt1 = mesh_proto.TrafficTrace{
+		Selectors: []*mesh_proto.Selector{{
+			Match: map[string]string{"serivce": "*"},
+		}},
+		Conf: &mesh_proto.TrafficTrace_Conf{
+			Backend: "tracing-backend",
+		},
+	}
+	pt1 = mesh_proto.ProxyTemplate{
+		Selectors: []*mesh_proto.Selector{{
+			Match: map[string]string{"serivce": "*"},
+		}},
+		Conf: &mesh_proto.ProxyTemplate_Conf{
+			Imports: []string{"default-kuma-profile"},
+		},
+	}
+)
+
+func kumaResources(response *v2.DiscoveryResponse) (resources []*mesh_proto.KumaResource, _ error) {
+	for _, r := range response.Resources {
+		kr := &mesh_proto.KumaResource{}
+		if err := ptypes.UnmarshalAny(r, kr); err != nil {
+			return nil, err
+		}
+		resources = append(resources, kr)
+	}
+	return
+}
+
+var _ = Describe("KDS Server", func() {
+	createServer := func(rt runtime.Runtime) kds_server.Server {
+		log := core.Log
+		hasher, cache := mads_server.NewXdsContext(log)
+		generator := kds_server.NewSnapshotGenerator(rt)
+		versioner := mads_server.NewVersioner()
+		reconciler := mads_server.NewReconciler(hasher, cache, generator, versioner)
+		syncTracker := mads_server.NewSyncTracker(reconciler, rt.Config().KdsServer.RefreshInterval)
+		callbacks := util_xds.CallbacksChain{
+			util_xds.LoggingCallbacks{Log: log},
+			syncTracker,
+		}
+		return kds_server.NewServer(cache, callbacks, log)
+	}
+
+	var tc testContext
+	BeforeEach(func() {
+		s := memory.NewStore()
+		mgr := manager.NewResourceManager(s)
+		srv := createServer(&testRuntimeContext{
+			rom: mgr,
+			cfg: kuma_cp.Config{
+				KdsServer: &kds_config.KumaDiscoveryServerConfig{
+					RefreshInterval: 1 * time.Second,
+				},
+			},
+		})
+		stream := makeMockStream()
+		stop := make(chan struct{})
+		go func() {
+			err := srv.StreamKumaResources(stream)
+			Expect(err).ToNot(HaveOccurred())
+			close(stop)
+		}()
+
+		tc = &testContextImpl{
+			resourceStore: s,
+			mockStream:    stream,
+			stopCh:        stop,
+			responses:     map[string]*v2.DiscoveryResponse{},
+		}
+	})
+
+	It("should support all existing resource types", func() {
```

Comment:
Is there a way to verify against `kds.SupportedTypes` that we really test "all" of them

Reply:
I've added a check which is quite useless and stupid, but it might remind to update these tests:
```go
Expect([]proto.Message{&mesh1, &ingress, &cb1, &fi1, &hc1, &tl1, &tp1, &tr1, &tt1, &pt1, &s1}).To(HaveLen(len(kds.SupportedTypes)))
```